### PR TITLE
vsearch based on hnsw (nmslib)

### PR DESCRIPTION
**Pull request by MXueguang**
current process to replicate numbers in ICTIR paper:
1. download bm25 hnsw index
```
wget https://www.dropbox.com/s/3i5vjhy4snj0yd8/bm25_hnsw_40_2500.tar.gz
tar -xvf /bm25_hnsw_40_2500.tar.gz
```
2. download msmarco-passage-dev-subset bm25 vectors:
```
wget https://www.dropbox.com/s/gkgeocv2ktfapl6/msmarco-passage-dev-subset-bm25-vectors.json
```
3. run vsearch
```
python -m pyserini.vsearch --index sparse_bm25_indexes/40_2500/ --topics msmarco-passage-dev-subset-bm25-vectors.json --output bm25_ann.run --is-sparse --dim 202514 --batch-size 36 --threads 12
```

Notes:
- we will need a module `/pyserini/vindex` to generate (1)
- (2) currently has following format
```
[
    {
        "id": 1102330,
        "topic": "why do people grind teeth in sleep",
        "vector": [
            [71919,  92561, 144918, 169850, 179980, 196713],
            [1.0, 1.0, 1.0, 1.0, 1.0, 1.0]
        ]
...
```

### CODE REVIEW

**Comment by MXueguang** [comment]
@lintool any suggestions?

**Comment by lintool** [comment]
hey @MXueguang this is great. Were you looking for a CR?

How's the effectiveness/efficiency (time and space) compared to Lucene?

**Comment by MXueguang** [comment]
space:
- Lucene: 2.6G
- HNSW: 4.2G
- 60% space increase

Effectiveness:
- Lucene (tuned): `recall_100` is 0.670
- HNSW: `recall_100` is 0.601
- 10% effectiveness drop

query-by-query search
- Lucene: 16.06 query/sec
- HNSW: 69 query/sec
- 330% efficiency increase

batch search, 12 threads
- Lucene: avg 63 query/sec
- HNSW: avg 200 query/sec 
- 220% efficiency increase

**Comment by lintool** [comment]
Hrm, so worse in every single way. Interesting...

**Comment by lintool** [comment]
Does dense work? Can you try to materialize vectors from any of our dense transformer models, index using nmslib's hnsw, and compare against faiss?

**Comment by MXueguang** [comment]
sure, I'll try with embedding from TCT_ColBERT

**Comment by MXueguang** [comment]
Faiss's hnsw v.s. nmslib's hnsw

space:
- Faiss: 43G
- nmslib: 69G

Effectiveness:
- Faiss: recall_100 is 0.8511
- nmslib: recall_100 is 0.8586

query-by-query search
- Faiss:  20.5 query/sec
- nmslib: 70 query/sec


batch search, 12 threads
- Faiss: avg 107 query/sec
- nmslib: avg 150 query/sec

File: pyserini/vsearch/__init__.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#
+# Pyserini: Python interface to the Anserini IR toolkit built on Lucene
```
Comment (lintool) [inline pyserini/vsearch/__init__.py:2]: Change no new boilerplate? (everywhere below also.)

File: pyserini/vsearch/_vsearcher.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import os
+from dataclasses import dataclass
+from typing import Dict, List
+
+import nmslib
```
Comment (lintool) [inline pyserini/vsearch/_vsearcher.py:21]: we should note the version we're using in #599 - also, add to requirements?

**Review by lintool — APPROVED** [review]

**Comment by lintool** [comment]
IIRC, hnsw indexing is very memory intensive? Can you document on this thread how much memory is required for the above experiments?

Also, I assume pre-built and integration tests are coming?

**Comment by JMMackenzie** [comment]
> space:
> 
>     * Lucene: 2.6G
> 
>     * HNSW: 4.2G
> 
>     * 60% space increase
> 
> 
> Effectiveness:
> 
>     * Lucene (tuned): `recall_100` is 0.670
> 
>     * HNSW: `recall_100` is 0.601
> 
>     * 10% effectiveness drop
> 
> 
> query-by-query search
> 
>     * Lucene: 16.06 query/sec
> 
>     * HNSW: 69 query/sec
> 
>     * 330% efficiency increase
> 
> 
> batch search, 12 threads
> 
>     * Lucene: avg 63 query/sec
> 
>     * HNSW: avg 200 query/sec
> 
>     * 220% efficiency increase
> 
> 
> Note:
> Lucene here is Pyserini-BM25
> HNSW here is nmslib-BM25

Hi @MXueguang, I have been following this PR and discussing with @lintool .

Just wondering, the numbers you've computed above for Lucene -- Do they correspond to the "default" MSMARCO passages run (AKA here: https://github.com/castorini/anserini/blob/master/docs/experiments-msmarco-passage.md )? Thanks!

**Comment by MXueguang** [comment]
@JMMackenzie yes, but here I am calculating Recall_**100**

**Comment by MXueguang** [comment]
Run code below to do hnsw sparse search for deepimpact
```
python -m pyserini.vsearch --index ${DATA_DIR}/hnsw_index/ \
                           --topics ${DATA_DIR}/deepimpact_msmarco_passage_dev_topic.jsonl \
                           --output ${DATA_DIR}/hnsw_40_2500_1000.run \
                           --ef 1000 \
                           --is-sparse 
```

**Pull request merged**